lib/get-eslint-modules.js

```diff
@@ -0,0 +1,44 @@
+var Module = require("module");
+var path = require("path");
+
+/**
+ * We can't just require in libraries related to ESLint, scope traversal etc.
+ * We need to look them up based on what the end user has installed.
+ *
+ * This function handles tracking down all the relevant libraries in this area
+ * and returns an object which includes them as properties.
+ */
+module.exports = function getESLintModules() {
+  try {
+    // avoid importing a local copy of eslint, try to find a peer dependency
+    var eslintLoc = Module._resolveFilename("eslint", module.parent);
+  } catch (err) {
+    try {
+      // avoids breaking in jest where module.parent is undefined
+      eslintLoc = require.resolve("eslint");
+    } catch (err) {
+      throw new ReferenceError("couldn't resolve eslint");
+    }
+  }
+
+  // get modules relative to what eslint will load
+  var eslintMod = new Module(eslintLoc);
+  eslintMod.filename = eslintLoc;
+  eslintMod.paths = Module._nodeModulePaths(path.dirname(eslintLoc));
+
+  var eslintScope = eslintMod.require("eslint-scope");
+  var Definition = eslintMod.require("eslint-scope/lib/definition").Definition;
+  var Referencer = eslintMod.require("eslint-scope/lib/referencer");
+  var Traverser = eslintMod.require("eslint/lib/util/traverser");
+  var ScopeManager = eslintMod.require("eslint-scope/lib/scope-manager");
```

Comment:
All of these APIs are internal, and are subject to change/be removed in the future. Is it necessary to import the specific modules that are being used by ESLint? Could `babel-eslint` use more generic modules instead, by depending on `eslint-scope` and `estraverse` directly? If not, what APIs would ESLint need to provide in order to make this possible?

Ideally, I think `babel-eslint` should only use ESLint's public APIs. As currently implemented, this will break if e.g. ESLint stops depending on `eslint-scope` in the future, or changes how `Traverser` works internally. The overarching goal here is to fix the recurring issue of "`babel-eslint` is broken due to an internal change in ESLint", but we can only achieve that if `babel-eslint` avoids using internal APIs (and ESLint provides sufficient public APIs to make everything work and be stable).